tests/suite/upgrade_test.go

```diff
@@ -0,0 +1,246 @@
+				},
+			},
+		}
+
+		for _, test := range tests {
+			go func(cfg testCfg) {
+				defer GinkgoRecover()
+
+				results, metrics := framework.RunLoadTest(
+					[]framework.Target{cfg.target},
+					100,
+					60*time.Second,
+					cfg.desc,
+					fmt.Sprintf("%s:%s", address, cfg.port),
+					"cafe.example.com",
+				)
+
+				scheme := strings.Split(cfg.target.URL, "://")[0]
+				metricsRes := metricsResults{
+					metrics:  &metrics,
+					testName: fmt.Sprintf("\n## Test: %s\n\n```text\n", cfg.desc),
+					scheme:   scheme,
+				}
+
+				buf := new(bytes.Buffer)
+				encoder := vegeta.NewCSVEncoder(buf)
+				for _, res := range results {
+					res := res
+					Expect(encoder.Encode(&res)).To(Succeed())
+				}
+
+				csvName := fmt.Sprintf("%s.csv", scheme)
+				filename := filepath.Join(resultsDir, csvName)
+				csvFile, err := framework.CreateResultsFile(filename)
+				Expect(err).ToNot(HaveOccurred())
+
+				_, err = fmt.Fprint(csvFile, buf.String())
+				Expect(err).ToNot(HaveOccurred())
+				csvFile.Close()
+
+				output, err := framework.GeneratePNG(resultsDir, csvName, fmt.Sprintf("%s.png", scheme))
+				Expect(err).ToNot(HaveOccurred(), string(output))
+
+				metricsCh <- &metricsRes
+			}(test)
+		}
+
+		// allow traffic flow to start
+		time.Sleep(2 * time.Second)
+
+		output, err := framework.UpgradeNGF(cfg, "--values", valuesFile)
+		Expect(err).ToNot(HaveOccurred(), string(output))
+
+		Expect(resourceManager.ApplyFromFiles([]string{"ngf-upgrade/gateway-updated.yaml"}, ns.Name)).To(Succeed())
+
+		podNames, err := framework.GetNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+
+		// ensure that the leader election lease has been updated to the new pods
+		var lease coordination.Lease
+		key := types.NamespacedName{Name: "ngf-test-nginx-gateway-fabric-leader-election", Namespace: ngfNamespace}
+		Expect(k8sClient.Get(ctx, key, &lease)).To(Succeed())
+
+		Expect(lease.Spec.HolderIdentity).ToNot(BeNil())
+		Expect(podNames).To(ContainElement(*lease.Spec.HolderIdentity))
+
+		// ensure that the Gateway has been properly updated with a new listener
+		var gw v1.Gateway
+		key = types.NamespacedName{Name: "gateway", Namespace: ns.Name}
+		Expect(wait.PollUntilContextCancel(
+			ctx,
+			500*time.Millisecond,
+			true, /* poll immediately */
+			func(ctx context.Context) (bool, error) {
+				Expect(k8sClient.Get(ctx, key, &gw)).To(Succeed())
+				expListenerName := "http-new"
+				for _, listener := range gw.Status.Listeners {
+					if listener.Name == v1.SectionName(expListenerName) {
+						return true, nil
+					}
+				}
+				return false, nil
+			},
+		)).To(Succeed())
+
+		recvdMetrics := make([]metricsResults, 0, 2)
+		// loop until we've received results from both tests, or timeout
+		for received := 0; received < 2; received++ {
```

Comment:
A wait group might be better. You can just wait for the group to finish and then range over the metricsResults channel.

Reply:
I did that originally, but felt that it was redundant if I was already waiting on the channel. Didn't seem to offer much value.

Reply:
I also would've had to build another goroutine and channel to allow for the waitgroup to timeout so I wasn't waiting forever, so this just felt cleaner.

Reply:
Do we need a timeout, though? Is there a case where the test can hang forever?

Reply:
Maybe not.

Reply:
If we still have to loop over the channel for 2 results though, does it make sense to use the wait group?

Reply:
I think so, because you need to know when to close the channel otherwise the range loop will hang. 

```
wg.Wait()
close(metricsCh)

range res := metricsCh{}
```

Reply:
Ok cool, didn't actually realize you can range over a closed channel.